### PR TITLE
S4-188: Added pending payment email logic

### PR DESCRIPTION
**Pull request by faggc**
# Description
Added code to send email requesting payment when dissolution status is set to "pending payment"

## Coding Standards

Code Style Guide: https://github.com/companieshouse/dissolution-web/wiki/Code-Style-Guide

## Checklist

- [x] 3 Amigos
- [x] Acceptance Criteria met
- [x] Branch named {feature|hotfix|task}/{S4-*}
- [x] Commit messages are meaningful
- [x] Manually tested
- [x] All unit tests passing
- [ ] API (Karate) tests passing
- [x] Pulled latest master into feature branch
- [x] Code reviewed
- [ ] New Docker environment variables have also been added to the revel1 environment and cidev environment
- [ ] If your pull request depends on any other, please link them in the description

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/mapper/email/DissolutionEmailMapper.java [hunk 1]
```diff
@@ -81,4 +83,23 @@ public SignatoryToSignEmailData mapToSignatoryToSignEmailData(Dissolution dissol
 
         return signatoryToSignEmailData;
     }
+
+    public PendingPaymentEmailData mapToPendingPaymentEmailData(Dissolution dissolution) {
+        final String applicantEmailAddress = dissolution.getCreatedBy().getEmail();
+        final String dissolutionReferenceNumber = dissolution.getData().getApplication().getReference();
+        final String companyNumber = dissolution.getCompany().getNumber();
+        final String companyName = dissolution.getCompany().getName();
```
Comment (peadarkelly) [inline src/main/java/uk/gov/companieshouse/mapper/email/DissolutionEmailMapper.java:91]: I don't think there is any need to make these separate vaiables. Could save around 5 lines of code here

File: src/main/java/uk/gov/companieshouse/model/dto/email/PendingPaymentEmailData.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
+package uk.gov.companieshouse.model.dto.email;
+
+public class PendingPaymentEmailData extends EmailData {
+    private String to;
+    private String subject;
```
Comment (peadarkelly) [inline src/main/java/uk/gov/companieshouse/model/dto/email/PendingPaymentEmailData.java:5]: Please rebase master - `to` and `subject` have been moved into the superclass and do not need to be redeclared here, along with their getters and setters

File: src/test/java/uk/gov/companieshouse/mapper/email/DissolutionEmailMapperTest.java [hunk 3]
```diff
@@ -89,6 +90,24 @@ public void mapToApplicationRejectedEmailData_mapsApplicationRejectedEmailData()
...
+
+        final PendingPaymentEmailData result = dissolutionEmailMapper.mapToPendingPaymentEmailData(dissolution);
+
+        assertEquals(pendingPaymentEmailData.getTo(), result.getTo());
+        assertEquals(pendingPaymentEmailData.getSubject(), result.getSubject());
+        assertEquals(pendingPaymentEmailData.getDissolutionReferenceNumber(), result.getDissolutionReferenceNumber());
+        assertEquals(pendingPaymentEmailData.getCompanyNumber(), result.getCompanyNumber());
+        assertEquals(pendingPaymentEmailData.getCompanyName(), result.getCompanyName());
+        assertEquals(pendingPaymentEmailData.getChsUrl(), result.getChsUrl());
+        assertEquals(pendingPaymentEmailData.getCdnHost(), result.getCdnHost());
```
Comment (peadarkelly) [inline src/test/java/uk/gov/companieshouse/mapper/email/DissolutionEmailMapperTest.java:108]: I don;t like these assertions. You are basically copying the implementation of the mapper. You are also relying on the fixtures containing valid data. This unit test would continue to pass if the fixtures was full of null values. Instead follow the example set by the mapToSignatoryToSignEmailData tests and hardcode the data you are asserting on.

File: src/main/java/uk/gov/companieshouse/service/dissolution/DissolutionEmailService.java [hunk 4]
```diff
@@ -63,6 +65,17 @@ public void sendApplicationOutcomeEmail(Dissolution dissolution, DissolutionVerd
         sendEmail(emailDocument);
     }
 
+    public void sendPendingPaymentEmail(Dissolution dissolution) {
```
Comment (peadarkelly) [inline src/main/java/uk/gov/companieshouse/service/dissolution/DissolutionEmailService.java:68]: Missing unit test for this

File: src/main/java/uk/gov/companieshouse/service/dissolution/DissolutionPatcher.java [hunk 5]
```diff
@@ -56,6 +56,7 @@ public DissolutionPatchResponse addDirectorApproval(String companyNumber, String
         if (!this.hasDirectorsLeftToApprove(dissolution)) {
             setDissolutionStatus(dissolution, ApplicationStatus.PENDING_PAYMENT);
             dissolution.setCertificate(this.certificateGenerator.generateDissolutionCertificate(dissolution));
+            dissolutionEmailService.sendPendingPaymentEmail(dissolution);
```
Comment (peadarkelly) [inline src/main/java/uk/gov/companieshouse/service/dissolution/DissolutionPatcher.java:59]: Missing unit test for this.

Comment (peadarkelly) [inline src/main/java/uk/gov/companieshouse/service/dissolution/DissolutionPatcher.java:59]: Could we move these 3 lines into a private method i.e. `handleFinalApproval(dissolution);` to keep the public method as light as possible

**Review by peadarkelly — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/companieshouse/service/dissolution/DissolutionEmailService.java (hunk 4 above)
Comment (sapphireduffy) [reply src/main/java/uk/gov/companieshouse/service/dissolution/DissolutionEmailService.java:68]: @faggc If you need any help with unit tests this let me know

**Review by sapphireduffy — COMMENTED** [review]
Looks great! Once we add Peadars suggestions I will approve! :D

File: src/test/java/uk/gov/companieshouse/mapper/email/DissolutionEmailMapperTest.java [hunk 6]
```diff
@@ -89,6 +91,56 @@ public void mapToApplicationRejectedEmailData_mapsApplicationRejectedEmailData()
         assertEquals(applicationRejectedEmailData.getRejectReasons(), result.getRejectReasons());
     }
 
+    @Test
```
Comment (peadarkelly) [inline src/test/java/uk/gov/companieshouse/mapper/email/DissolutionEmailMapperTest.java:94]: This test can be removed now that the following 2 tests cover everything

**Review by peadarkelly — APPROVED** [review]
one final comment

**Pull request merged**